src/std/misc/evector.ss

```diff
@@ -0,0 +1,321 @@
+;;; -*- Gerbil -*-
+;;; © fare, vyzo
+(export make-evector evector?
+        make-ebytes ebytes?
+        make-ebits ebits?
+        (rename: evector-set-fill-pointer! evector-fill-pointer-set!)
+        evector-fill-pointer
+        (rename: ebytes-set-fill-pointer! ebytes-fill-pointer-set!)
+        ebytes-fill-pointer
+        (rename: ebits-set-fill-pointer! ebits-fill-pointer-set!)
+        ebits-fill-pointer)
+
+;; Simple extensible vectors, byte-vectors, and bit-vectors
+;; They double their size each time they need to grow,
+;; or you can insist on controlling what size they grow to.
+
+(import
+  :gerbil/gambit
+  :std/srfi/43
+  :std/error
+  :std/sugar
+  ./bytes
+  ./number)
+
+(declare (not safe))
+
+(defstruct evector (vector fill-pointer)
+  transparent: #t unchecked: #t
+  constructor: :init!)
+
+(defstruct ebytes (bytes fill-pointer)
+  transparent: #t unchecked: #t
+  constructor: :init!)
+
+(defstruct ebits (bits fill-pointer)
+  transparent: #t unchecked: #t
+  constructor: :init!)
+
+(defmethod {:init! evector}
+  (lambda (self vector fp)
+    (check-argument-vector vector)
+    (check-argument-fxlength fp)
+    (struct-instance-init! self vector fp)))
+
+(defmethod {:init! ebytes}
+  (lambda (self bytes fp)
+    (check-argument-u8vector bytes)
+    (check-argument-fxlength fp)
+    (struct-instance-init! self bytes fp)))
+
+(defmethod {:init! ebits}
+  (lambda (self bytes fp)
+    (check-argument-u8vector bytes)
+    (check-argument-fxlength fp)
+    (struct-instance-init! self bytes fp)))
+
+(defcheck-argument-type evector)
+(defcheck-argument-type ebytes)
+(defcheck-argument-type ebits)
+
+;; TODO these could be moved to a stdlib module, say :std/typecheck
+(def (fxlength? o)
+  (and (fixnum? o) (fx>= o 0)))
+
+(defcheck-argument-type fixnum)
```

Comment:
the def-check-argument-type macro I had defined in sugar accepts multiple types as arguments.

Reply:
it still does, but I prefer to do a single thing at a time